src/wp-includes/abilities-api/class-wp-abilities-category-registry.php

```diff
@@ -0,0 +1,247 @@
+<?php
+/**
+ * Abilities API
+ *
+ * Defines WP_Abilities_Category_Registry class.
+ *
+ * @package WordPress
+ * @subpackage Abilities API
+ * @since 6.9.0
+ */
+
+declare( strict_types = 1 );
+
+/**
+ * Manages the registration and lookup of ability categories.
+ *
+ * @since 6.9.0
+ * @access private
+ */
+final class WP_Abilities_Category_Registry {
+	/**
+	 * The singleton instance of the registry.
+	 *
+	 * @since 6.9.0
+	 * @var ?self
+	 */
+	private static $instance = null;
+
+	/**
+	 * Holds the registered categories.
+	 *
+	 * @since 6.9.0
+	 * @var \WP_Ability_Category[]
+	 */
+	private $registered_categories = array();
+
+	/**
+	 * Registers a new category.
+	 *
+	 * Do not use this method directly. Instead, use the `wp_register_ability_category()` function.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see wp_register_ability_category()
+	 *
+	 * @param string              $slug The unique slug for the category. Must contain only lowercase
+	 *                                  alphanumeric characters and dashes.
+	 * @param array<string,mixed> $args An associative array of arguments for the category. See wp_register_ability_category() for
+	 *                                  details.
+	 * @return ?\WP_Ability_Category The registered category instance on success, null on failure.
+	 *
+ * @phpstan-param array{
+ *   label: string,
+ *   description: string,
+ *   meta?: array<string,mixed>,
+ *   ...<string, mixed>
+ * } $args
+	 */
+	public function register( string $slug, array $args ): ?WP_Ability_Category {
+		if ( ! doing_action( 'abilities_api_categories_init' ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				sprintf(
+					/* translators: 1: abilities_api_categories_init, 2: category slug. */
+					esc_html__( 'Categories must be registered during the %1$s action. The category %2$s was not registered.' ),
+					'<code>abilities_api_categories_init</code>',
+					'<code>' . esc_html( $slug ) . '</code>'
+				),
+				'6.9.0'
+			);
+			return null;
+		}
+
+		if ( $this->is_registered( $slug ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: Category slug. */
+				esc_html( sprintf( __( 'Category "%s" is already registered.' ), $slug ) ),
```

Comment:
```suggestion
				sprintf( __( 'Category "%s" is already registered.' ), esc_html( $slug ) ),
```

The entire string does not need to be escaped